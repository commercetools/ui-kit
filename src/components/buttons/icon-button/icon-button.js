import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import { compose } from 'recompose';
import invariant from 'tiny-invariant';
import isNil from 'lodash.isnil';
import withMouseDownState from '../../../hocs/with-mouse-down-state';
import withMouseOverState from '../../../hocs/with-mouse-over-state';
import filterDataAttributes from '../../../utils/filter-data-attributes';
import AccessibleButton from '../accessible-button';
import styles from './icon-button.mod.css';

const getStateClassNames = (isDisabled, isToggled) =>
  classnames({
    [styles.default]: !isToggled && !isDisabled,
    [styles.active]: isToggled,
    [styles.disabled]: isDisabled,
  });

// Gets the color which the icon should have based on context of button's state/cursor behavior
const getIconThemeColor = props => {
  // if button has a theme, icon should be white when hovering/clicking
  if (props.theme !== 'default' && (props.isToggled || props.isMouseOver)) {
    if (props.isDisabled) {
      return 'grey';
    }
    return 'white';
  }

  // if button is disabled, icon should be grey
  if (props.isDisabled) return 'grey';
  // if button is not disabled nor has a theme, return icon's default color
  return props.icon.props.theme;
};

const getShapeClassName = shape => styles[shape];
const getSizeClassName = size => styles[`container-${size}`];
const getThemeClassName = theme => {
  if (!theme) return undefined;

  const themeClassName = styles[`theme-${theme}`];

  // Whenever a theme is specified a fitting theme className should
  // be available except for the default theme.
  invariant(
    themeClassName || theme === 'default',
    `ui-kit/IconButton: the specified theme '${theme}' is not supported.`
  );

  return themeClassName;
};

export const IconButton = props => {
  const dataProps = {
    'data-track-component': 'IconButton',
    ...filterDataAttributes(props),
  };
  return (
    <div
      onMouseDown={props.handleMouseDown}
      onMouseUp={props.handleMouseUp}
      onMouseOver={props.handleMouseOver}
      onMouseOut={props.handleMouseOut}
      className={classnames(
        styles['button-appearance'],
        getStateClassNames(props.isDisabled, props.isToggled),
        getShapeClassName(props.shape),
        getSizeClassName(props.size),
        getThemeClassName(props.theme)
      )}
    >
      <AccessibleButton
        buttonAttributes={dataProps}
        label={props.label}
        onClick={props.onClick}
        isToggled={props.isToggled}
        isDisabled={props.isDisabled}
        className={styles[`container-${props.size}`]}
      >
        <div
          className={classnames(
            styles['button-layout'],
            styles[`container-${props.size}`]
          )}
        >
          {props.icon &&
            React.cloneElement(props.icon, {
              size: props.size,
              theme: getIconThemeColor(props),
            })}
        </div>
      </AccessibleButton>
    </div>
  );
};

IconButton.propTypes = {
  label: PropTypes.string.isRequired,
  icon: PropTypes.node,
  isToggleButton: PropTypes.bool.isRequired,
  isToggled(props, propName, componentName, ...rest) {
    if (props.isToggleButton) {
      return PropTypes.bool.isRequired(props, propName, componentName, ...rest);
    }
    if (!isNil(props[propName]))
      return new Error(
        `Invalid prop \`${propName}\` supplied to \`${componentName}\`. \`${propName}\` does not have any effect when the button is not a toggle button.`
      );
    return PropTypes.bool(props, propName, componentName, ...rest);
  },
  isDisabled: PropTypes.bool,
  /* FIXME: onClick should be required.
    There are still some places where we can't pass it yet.
    Check the spreadsheet to see where;
  */
  onClick: PropTypes.func,
  shape: PropTypes.oneOf(['round', 'square']),
  size: PropTypes.oneOf(['small', 'medium', 'big']),
  theme(props, propName, componentName, ...rest) {
    if (props[propName] !== 'default' && !props.isToggleButton) {
      return new Error(
        `Invalid prop \`${propName}\` supplied to \`${componentName}\`. Only toggle buttons may have a theme.`
      );
    }
    return PropTypes.oneOf(['default', 'green', 'blue'])(
      props,
      propName,
      componentName,
      ...rest
    );
  },

  // HoC
  isMouseDown: PropTypes.bool.isRequired,
  isMouseOver: PropTypes.bool.isRequired,
  handleMouseOver: PropTypes.func.isRequired,
  handleMouseOut: PropTypes.func.isRequired,
  handleMouseDown: PropTypes.func.isRequired,
  handleMouseUp: PropTypes.func.isRequired,
};

IconButton.defaultProps = {
  theme: 'default',
  size: 'big',
  shape: 'round',
  isToggleButton: false,
};

IconButton.displayName = 'IconButton';

export default compose(
  withMouseOverState,
  withMouseDownState
)(IconButton);
