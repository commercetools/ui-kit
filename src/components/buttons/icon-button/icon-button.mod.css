@import '../../../../materials/borders.mod.css';
@import '../../../../materials/colors/base-colors.mod.css';
@import '../../../../materials/shadows/base-shadows.mod.css';
@import '../../../../materials/spacings.mod.css';
@import '../../../../materials/transitions.mod.css';

:root {
  --size-small-button: 16px;
  --size-medium-button: 24px;
  --size-big-button: 32px;
}

/*
* === BUTTON ELEMENT ===
*/

.button-appearance {
  display: inline-block;
  background-color: var(--color-white);
  box-shadow: var(--shadow-7-first), var(--shadow-7-second);
  color: var(--color-black);
  transition: background-color var(--transition-linear-80ms),
    box-shadow 150ms ease-in-out;
}

.button-layout {
  display: flex;
  align-items: center;
  justify-content: center;
}

.button-layout > * + * {
  margin: 0 0 0 6px;
}

/*
* === CONTAINER ===
*/

/*
* — Shapes —
*/

.round {
  border-radius: 50%;
}

.square.container-small {
  border-radius: var(--border-radius-2);
}
.square.container-medium {
  border-radius: var(--border-radius-4);
}
.square.container-big {
  border-radius: var(--border-radius-6);
}

/*
* — Sizes —
*/

.container-small {
  height: var(--size-small-button);
  width: var(--size-small-button);
}

.container-medium {
  height: var(--size-medium-button);
  width: var(--size-medium-button);
}

.container-big {
  height: var(--size-big-button);
  width: var(--size-big-button);
}

/*
* === THEMES ===
*/

/*
* — Default —
*/

/* Disabled button */

.disabled {
  background-color: var(--color-navy-98);
  color: var(--color-gray-60);
  box-shadow: none;
}

/* Default button */

.default:hover {
  box-shadow: var(--shadow-8-first), var(--shadow-8-second);
}

.default:active {
  box-shadow: var(--shadow-9-first), var(--shadow-9-second),
    var(--shadow-9-third);
  background-color: var(--color-white);
}

/* Toggled button */

.active {
  box-shadow: var(--shadow-9-first), var(--shadow-9-second),
    var(--shadow-9-third);
  background-color: var(--color-white);
}

.active:hover {
  box-shadow: var(--shadow-9-first), var(--shadow-9-second),
    var(--shadow-9-third);
  background-color: var(--color-gray-95);
}

/*
* — Blue Theme —
*/

/* Disabled-Active button */

.disabled.theme-blue:hover {
  background-color: var(--color-blue-85);
}

/* When the button is active (isToggled: true) and somehow is disabled
  it should have a different color to indicate that it's active but can't
  receive any actions */

.active.disabled.theme-blue {
  background-color: var(--color-blue-85);
  color: var(--color-white);
  box-shadow: var(--shadow-9-first), var(--shadow-9-second),
    var(--shadow-9-third);
}

/* Toggled button */

.theme-blue.active,
.theme-blue:hover {
  background-color: var(--color-blue);
  color: var(--color-white);
}

.active.theme-blue:hover {
  background-color: var(--color-blue-85);
}

/*
* — Green Theme —
*/

/* Disabled-Active button */

/* When the button is active (isToggled: true) and somehow is disabled
  it should have a different color to indicate that it's active but can't
  receive any actions */

.disabled.theme-green:hover {
  background-color: var(--color-green-85);
}

.active.disabled.theme-green {
  background-color: var(--color-green-85);
  color: var(--color-white);
  box-shadow: var(--shadow-9-first), var(--shadow-9-second),
    var(--shadow-9-third);
}

/* Toggled button */

.theme-green.active,
.theme-green:hover {
  background-color: var(--color-green);
  color: var(--color-white);
}

.active.theme-green:hover {
  background-color: var(--color-green-85);
}
