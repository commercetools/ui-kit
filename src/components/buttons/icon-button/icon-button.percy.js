import React from 'react';
import IconButton from './icon-button';
import { InformationIcon } from '../../icons';

suite('IconButton', () => {
  percySnapshot('regular', () => (
    <IconButton
      icon={<InformationIcon />}
      label="A label text"
      onClick={() => {}}
    />
  ));

  percySnapshot('disabled', () => (
    <IconButton
      icon={<InformationIcon />}
      label="A label text"
      onClick={() => {}}
      isDisabled={true}
    />
  ));

  suite('as toggle button', () => {
    percySnapshot('when toggled', () => (
      <IconButton
        icon={<InformationIcon />}
        label="A label text"
        onClick={() => {}}
        isToggleButton={true}
        isToggled={true}
      />
    ));

    percySnapshot('when not toggled', () => (
      <IconButton
        icon={<InformationIcon />}
        label="A label text"
        onClick={() => {}}
        isToggleButton={true}
      />
    ));
  });

  suite('shapes', () => {
    percySnapshot('when round (default)', () => (
      <IconButton
        icon={<InformationIcon />}
        shape="round"
        label="A label text"
        onClick={() => {}}
      />
    ));

    percySnapshot('when square', () => (
      <IconButton
        icon={<InformationIcon />}
        shape="square"
        label="A label text"
        onClick={() => {}}
      />
    ));
  });

  suite('sizes', () => {
    suite('when round', () => {
      percySnapshot('when small', () => (
        <IconButton
          icon={<InformationIcon />}
          size="small"
          label="A label text"
          onClick={() => {}}
        />
      ));

      percySnapshot('when medium', () => (
        <IconButton
          icon={<InformationIcon />}
          size="medium"
          label="A label text"
          onClick={() => {}}
        />
      ));

      percySnapshot('when big (default)', () => (
        <IconButton
          icon={<InformationIcon />}
          size="big"
          label="A label text"
          onClick={() => {}}
        />
      ));
    });

    suite('when square', () => {
      percySnapshot('when small', () => (
        <IconButton
          icon={<InformationIcon />}
          shape="square"
          size="small"
          label="A label text"
          onClick={() => {}}
        />
      ));

      percySnapshot('when medium', () => (
        <IconButton
          icon={<InformationIcon />}
          shape="square"
          size="medium"
          label="A label text"
          onClick={() => {}}
        />
      ));

      percySnapshot('when big (default)', () => (
        <IconButton
          icon={<InformationIcon />}
          shape="square"
          size="big"
          label="A label text"
          onClick={() => {}}
        />
      ));
    });
  });

  suite('theme', () => {
    percySnapshot('when default', () => (
      <IconButton
        icon={<InformationIcon />}
        label="A label text"
        onClick={() => {}}
        isToggleButton={true}
        theme="default"
      />
    ));

    percySnapshot('when green', () => (
      <IconButton
        icon={<InformationIcon />}
        label="A label text"
        onClick={() => {}}
        isToggleButton={true}
        theme="green"
      />
    ));

    percySnapshot('when blue', () => (
      <IconButton
        icon={<InformationIcon />}
        label="A label text"
        onClick={() => {}}
        isToggleButton={true}
        theme="blue"
      />
    ));
  });
});
