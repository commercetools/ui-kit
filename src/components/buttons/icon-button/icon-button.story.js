import React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { withKnobs, boolean, select, text } from '@storybook/addon-knobs/react';
import Section from '../../../../.storybook/decorators/section';
import * as icons from '../../icons';
import Readme from './README.md';
import IconButton from './icon-button';

const iconNames = Object.keys(icons);

storiesOf('Components|Buttons', module)
  .addDecorator(withKnobs)
  .addParameters({
    readme: {
      // Show readme at the addons panel
      sidebar: Readme,
    },
  })
  .add('IconButton', () => (
    <Section>
      <IconButton
        type={select('type', ['submit', 'reset', 'button'], 'button')}
        shape={select('shape', ['round', 'square'], 'round')}
        size={select('size', ['big', 'medium', 'small'], 'big')}
        theme={select('theme', ['green', 'blue', 'default'], 'default')}
        icon={React.createElement(
          icons[select('icon', iconNames, iconNames[0])]
        )}
        onClick={action('onClick')}
        label={text('label', 'Accessibility text')}
        isToggleButton={boolean('isToggleButton', false)}
        isToggled={boolean('isToggled', false)}
        isDisabled={boolean('isDisabled?', false)}
      />
    </Section>
  ));
