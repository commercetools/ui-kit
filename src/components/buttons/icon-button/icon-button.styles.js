import invariant from 'tiny-invariant';
import { css } from '@emotion/core';
import vars from '../../../../materials/custom-properties';

const buttonSizes = {
  small: '16px',
  medium: '24px',
  big: '32px',
};

const getStateStyles = (isDisabled, isActive, theme) => {
  if (isDisabled) {
    const disabledStyle = css`
      background-color: ${vars.colorAccent98};
      border-color: ${vars.colorNeutral};
      color: ${vars.colorNeutral60};
      box-shadow: none;
    `;
    switch (theme) {
      case 'info':
        return [
          disabledStyle,
          css`
            &:hover {
              border-color: ${vars.colorInfo85};
              background-color: ${vars.colorInfo85};
            }
          `,
        ];
      case 'primary':
        return [
          disabledStyle,
          css`
            &:hover {
              border-color: ${vars.colorPrimary85};
              background-color: ${vars.colorPrimary85};
            }
          `,
        ];
      default:
        return disabledStyle;
    }
  }
  if (isActive) {
    const activeStyle = css`
      box-shadow: ${vars.shadow9};
      background-color: ${vars.colorSurface};
      border-color: ${vars.colorSurface};
      &:hover {
        box-shadow: ${vars.shadow9};
        background-color: ${vars.colorNeutral95};
        border-color: ${vars.colorNeutral95};
      }
    `;
    switch (theme) {
      case 'info':
        return [
          activeStyle,
          isDisabled &&
            // When the button is active and somehow is disabled it should have
            // a different color to indicate that it's active but can't receive any actions
            css`
              background-color: ${vars.colorInfo85};
              border-color: ${vars.colorInfo85};
              color: ${vars.colorSurface};
              box-shadow: ${vars.shadow9};
            `,
          css`
            background-color: ${vars.colorInfo};
            border-color: ${vars.colorInfo};
            color: ${vars.colorSurface};
            &:hover {
              background-color: ${vars.colorInfo85};
              border-color: ${vars.colorInfo85};
            }
          `,
        ];
      case 'primary':
        return [
          activeStyle,
          // When the button is active and somehow is disabled it should have
          // a different color to indicate that it's active but can't receive any actions
          isDisabled &&
            css`
              background-color: ${vars.colorPrimary85};
              border-color: ${vars.colorPrimary85};
              color: ${vars.colorSurface};
              box-shadow: ${vars.shadow9};
            `,

          css`
            background-color: ${vars.colorPrimary};
            color: ${vars.colorSurface};
            &:hover {
              background-color: ${vars.colorPrimary85};
              border-color: ${vars.colorPrimary85};
            }
          `,
        ];
      default:
        return activeStyle;
    }
  }
  return css`
    &:hover {
      box-shadow: ${vars.shadow8};
    }
    &:active {
      box-shadow: ${vars.shadow9};
      background-color: ${vars.colorSurface};
      border-color: ${vars.colorSurface};
    }
  `;
};

const getShapeStyles = (shape, size) => {
  switch (shape) {
    case 'round':
      return css`
        border-radius: 50%;
      `;
    case 'square':
      switch (size) {
        case 'small':
          return css`
            border-radius: ${vars.borderRadius2};
          `;
        case 'medium':
          return css`
            border-radius: ${vars.borderRadius4};
          `;
        case 'big':
          return css`
            border-radius: ${vars.borderRadius6};
          `;
        default:
          return css``;
      }
    default:
      return css``;
  }
};
const getSizeStyles = size => {
  switch (size) {
    case 'small':
      return css`
        height: ${buttonSizes.small};
        width: ${buttonSizes.small};
      `;
    case 'medium':
      return css`
        height: ${buttonSizes.medium};
        width: ${buttonSizes.medium};
      `;
    case 'big':
      return css`
        height: ${buttonSizes.big};
        width: ${buttonSizes.big};
      `;
    default:
      return css``;
  }
};
const getThemeStyles = theme => {
  if (!theme) return css``;

  if (theme === 'default') return css``;

  switch (theme) {
    case 'primary':
      return css`
        &:hover {
          background-color: ${vars.colorPrimary};
          border-color: ${vars.colorPrimary};
          color: ${vars.colorSurface};
        }
      `;
    case 'info':
      return css`
        &:hover {
          background-color: ${vars.colorInfo};
          border-color: ${vars.colorInfo};
          color: ${vars.colorSurface};
        }
      `;
    default: {
      invariant(
        false,
        `ui-kit/IconButton: the specified theme '${theme}' is not supported.`
      );
      return css``;
    }
  }
};

const getHoverStyles = (isDisabled, theme) => {
  if (theme === 'default' || isDisabled) return css``;

  return css`
    &:hover {
      * {
        fill: ${vars.colorSurface};
      }
    }
  `;
};

export {
  getStateStyles,
  getHoverStyles,
  getShapeStyles,
  getSizeStyles,
  getThemeStyles,
};
