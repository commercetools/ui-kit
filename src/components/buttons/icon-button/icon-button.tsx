import React from 'react';
import { compose } from 'recompose';
import { css } from '@emotion/core';
import vars from '../../../../materials/custom-properties';
import withMouseDownState from '../../../hocs/with-mouse-down-state';
import withMouseOverState from '../../../hocs/with-mouse-over-state';
import filterDataAttributes from '../../../utils/filter-data-attributes';
import AccessibleButton from '../accessible-button';
import invariant from 'tiny-invariant';
import {
  getStateStyles,
  getShapeStyles,
  getSizeStyles,
  getThemeStyles,
} from './icon-button.styles';

type IconButtonProps = {
  type: 'submit' | 'reset' | 'button';
  label: string;
  icon: React.ReactElement<any>;
  isToggleButton: boolean;
  isToggled: boolean;
  isDisabled?: boolean;
  onClick?: (e: React.MouseEvent<any>) => void;
  shape: 'round' | 'square';
  size: 'small' | 'medium' | 'big';
  theme: 'default' | 'green' | 'blue';

  // HoC
  isMouseDown: boolean;
  isMouseOver: () => void;
  handleMouseOver: () => void;
  handleMouseOut: () => void;
  handleMouseDown: () => void;
  handleMouseUp: () => void;
};

// Gets the color which the icon should have based on context of button's state/cursor behavior
const getIconThemeColor = (props: IconButtonProps) => {
  const isActive = props.isToggleButton && props.isToggled;
  // if button has a theme, icon should be white when hovering/clicking
  if (props.theme !== 'default' && (isActive || props.isMouseOver)) {
    if (props.isDisabled) {
      return 'grey';
    }
    return 'white';
  }

  // if button is disabled, icon should be grey
  if (props.isDisabled) return 'grey';
  // if button is not disabled nor has a theme, return icon's default color
  return props.icon.props.theme;
};

export const IconButton: React.FC<IconButtonProps> = props => {
  invariant(
    props.isToggleButton && typeof props.isToggled === 'undefined',
    'ui-kit/IconButton: when `isToggleButton` is passed `isToggled` is required.'
  );
  invariant(
    !props.isToggleButton && typeof props.isToggled !== 'undefined',
    'ui-kit/IconButton: when not `isToggleButton` `isToggled` does not affect the component.'
  );
  invariant(
    !props.isToggleButton && props.theme !== 'default',
    'ui-kit/IconButton: when not `isToggleButton` a `theme` is not supported.'
  );

  const buttonAttributes = {
    'data-track-component': 'IconButton',
    ...filterDataAttributes(props),
  };
  const isActive = props.isToggleButton && props.isToggled;
  return (
    <div
      onMouseDown={props.handleMouseDown}
      onMouseUp={props.handleMouseUp}
      onMouseOver={props.handleMouseOver}
      onMouseOut={props.handleMouseOut}
      css={[
        css`
          display: inline-block;
          background-color: ${vars.colorWhite};
          box-shadow: ${vars.shadow7};
          color: ${vars.colorBlack};
          transition: background-color ${vars.transitionLinear80Ms},
            box-shadow 150ms ease-in-out;
        `,
        getStateStyles(props.isDisabled, isActive, props.theme),
        getShapeStyles(props.shape, props.size),
        getSizeStyles(props.size),
        getThemeStyles(props.theme),
      ]}
    >
      <AccessibleButton
        buttonAttributes={buttonAttributes}
        type={props.type}
        label={props.label}
        onClick={props.onClick}
        isToggleButton={props.isToggleButton}
        isToggled={props.isToggled}
        isDisabled={props.isDisabled}
        css={getSizeStyles(props.size)}
      >
        <div
          css={[
            css`
              display: flex;
              align-items: center;
              justify-content: center;
            `,
            getSizeStyles(props.size),
          ]}
        >
          {props.icon &&
            React.cloneElement(props.icon, {
              size: props.size,
              theme: getIconThemeColor(props),
            })}
        </div>
      </AccessibleButton>
    </div>
  );
};

IconButton.defaultProps = {
  type: 'button',
  theme: 'default',
  size: 'big',
  shape: 'round',
  isToggleButton: false,
};

IconButton.displayName = 'IconButton';

export default compose(
  withMouseOverState,
  withMouseDownState
)(IconButton);
