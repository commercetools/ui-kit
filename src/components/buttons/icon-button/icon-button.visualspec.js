import { percySnapshot } from '@percy/puppeteer';

describe('IconButton', () => {
  beforeAll(async () => {
    await page.goto(`${HOST}/icon-button`);
  });

  it('Default', async () => {
    await expect(page).toMatch('A label text');
    await percySnapshot(page, 'IconButton');
  });
});
