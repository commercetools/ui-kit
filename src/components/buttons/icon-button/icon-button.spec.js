import React from 'react';
import { shallow } from 'enzyme';
import { FilterIcon } from '../../icons';
import { IconButton } from './icon-button';

const createProps = custom => ({
  type: 'button',
  label: '',
  onClick: () => {},
  icon: '',

  // HoC
  isMouseDown: false,
  isMouseOver: false,
  handleMouseUp: jest.fn(),
  handleMouseDown: jest.fn(),
  handleMouseOver: jest.fn(),
  handleMouseOut: jest.fn(),

  ...custom,
});

describe('rendering', () => {
  describe('structure', () => {
    let props;
    let wrapper;

    beforeEach(() => {
      props = createProps();
      wrapper = shallow(<IconButton {...props} />);
    });

    it('should output correct tree', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('should render a <AccessibleButton/>', () => {
      expect(wrapper).toRender('AccessibleButton');
    });

    it('should apply `button-layout` class name to `AccessibleButton`', () => {
      expect(wrapper.find('AccessibleButton')).toHaveClassName('container-big');
    });

    it('should supply `type` to `<AccessibleButton />`', () => {
      expect(wrapper.find('AccessibleButton')).toHaveProp('type', props.type);
    });

    it('should supply `onClick` to `<AccessibleButton />`', () => {
      expect(wrapper.find('AccessibleButton')).toHaveProp(
        'onClick',
        props.onClick
      );
    });

    it('should supply `isToggled` to `<AccessibleButton />`', () => {
      expect(wrapper.find('AccessibleButton')).toHaveProp(
        'isToggled',
        props.isToggled
      );
    });

    it('should supply `label` to `<AccessibleButton />`', () => {
      expect(wrapper.find('AccessibleButton')).toHaveProp('label', props.label);
    });

    describe('with data-* props', () => {
      beforeEach(() => {
        props = createProps({
          'data-track-component': 'IconButton',
          'data-track-label': 'IconButton',
          'data-track-event': 'click',
          'data-test': 'icon-button',
        });
        wrapper = shallow(<IconButton {...props} />);
      });
      it('should apply given `data-track-component` to AccessibleButton', () => {
        expect(wrapper.find('AccessibleButton')).toHaveProp(
          'buttonAttributes',
          expect.objectContaining({
            'data-track-component': 'IconButton',
          })
        );
      });
      it('should apply given `data-track-event` to AccessibleButton', () => {
        expect(wrapper.find('AccessibleButton')).toHaveProp(
          'buttonAttributes',
          expect.objectContaining({
            'data-track-event': 'click',
          })
        );
      });
      it('should apply given `data-track-label` to AccessibleButton', () => {
        expect(wrapper.find('AccessibleButton')).toHaveProp(
          'buttonAttributes',
          expect.objectContaining({
            'data-track-label': 'IconButton',
          })
        );
      });
      it('should apply given `data-test` to AccessibleButton', () => {
        expect(wrapper.find('AccessibleButton')).toHaveProp(
          'buttonAttributes',
          expect.objectContaining({
            'data-test': 'icon-button',
          })
        );
      });
    });
    describe('without data-* props', () => {
      beforeEach(() => {
        props = createProps();
        wrapper = shallow(<IconButton {...props} />);
      });
      it('should apply default `data-track-component` to AccessibleButton', () => {
        expect(wrapper.find('AccessibleButton')).toHaveProp(
          'buttonAttributes',
          expect.objectContaining({
            'data-track-component': 'IconButton',
          })
        );
      });
    });
  });

  describe('icon', () => {
    let props;
    let wrapper;
    beforeEach(() => {
      props = createProps({
        icon: <FilterIcon />,
      });
      wrapper = shallow(<IconButton {...props} />);
    });

    it('should output correct tree', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('should render an icon', () => {
      expect(wrapper).toRender(FilterIcon);
    });

    describe('when IconButton size is medium', () => {
      describe('when Icon size is big', () => {
        beforeEach(() => {
          props = createProps({
            size: 'medium',
            icon: <FilterIcon size="big" />,
          });
          wrapper = shallow(<IconButton {...props} />);
        });

        it('should overwrite the Icon size to be medium', () => {
          expect(wrapper.find(FilterIcon)).toHaveProp('size', 'medium');
        });
      });
    });

    describe('when theme is not default', () => {
      describe('when toggled', () => {
        beforeEach(() => {
          props = createProps({
            theme: 'blue',
            isToggled: true,
            isToggleButton: true,
            icon: <FilterIcon theme="red" />,
          });
          wrapper = shallow(<IconButton {...props} />);
        });

        it('should overwrite the Icon theme to be white', () => {
          expect(wrapper.find(FilterIcon)).toHaveProp('theme', 'white');
        });
      });

      describe('when mouse is down and mouse is over', () => {
        beforeEach(() => {
          props = createProps({
            theme: 'blue',
            isMouseDown: true,
            isMouseOver: true,
            isToggled: false,
            isToggleButton: true,
            icon: <FilterIcon theme="black" />,
          });
          wrapper = shallow(<IconButton {...props} />);
        });

        it('should overwrite the Icon theme to be white', () => {
          expect(wrapper.find(FilterIcon)).toHaveProp('theme', 'white');
        });
        it('should not overwrite the Icon theme to be black', () => {
          expect(wrapper.find(FilterIcon)).not.toHaveProp('theme', 'black');
        });
      });
      describe('when mouse is down and mouse is out', () => {
        beforeEach(() => {
          props = createProps({
            theme: 'blue',
            isMouseDown: true,
            isMouseOver: false,
            isToggled: false,
            isToggleButton: true,
            icon: <FilterIcon theme="black" />,
          });
          wrapper = shallow(<IconButton {...props} />);
        });

        it('should not overwrite the Icon theme to be white', () => {
          expect(wrapper.find(FilterIcon)).not.toHaveProp('theme', 'white');
        });
        it('should overwrite the Icon theme to be black', () => {
          expect(wrapper.find(FilterIcon)).toHaveProp('theme', 'black');
        });
      });
    });
  });

  describe('sizes', () => {
    describe('when small', () => {
      let wrapper;
      beforeEach(() => {
        const props = createProps({ size: 'small' });
        wrapper = shallow(<IconButton {...props} />);
      });

      it('should add the small class to container', () => {
        expect(wrapper).toHaveClassName('container-small');
      });
    });

    describe('when medium', () => {
      let wrapper;
      beforeEach(() => {
        const props = createProps({ size: 'medium' });
        wrapper = shallow(<IconButton {...props} />);
      });
      it('should add the medium class to container', () => {
        expect(wrapper).toHaveClassName('container-medium');
      });
    });

    describe('when big', () => {
      let wrapper;
      beforeEach(() => {
        const props = createProps({ size: 'big' });
        wrapper = shallow(<IconButton {...props} />);
      });
      it('should add the big class to container', () => {
        expect(wrapper).toHaveClassName('container-big');
      });
    });
  });

  describe('shapes', () => {
    describe('when no shape is provided', () => {
      let wrapper;
      beforeEach(() => {
        const props = createProps({ shape: undefined });
        wrapper = shallow(<IconButton {...props} />);
      });
      it('should add the round class', () => {
        expect(wrapper).toHaveClassName('default round container-big');
      });
    });
    describe('when providing shape', () => {
      let wrapper;
      beforeEach(() => {
        const props = createProps({ shape: 'square' });
        wrapper = shallow(<IconButton {...props} />);
      });
      it('should add the round class', () => {
        expect(wrapper).toHaveClassName('default square container-big');
      });
    });
  });

  describe('states', () => {
    describe('when is active', () => {
      let wrapper;
      beforeEach(() => {
        const props = createProps({
          isToggleButton: true,
          isToggled: true,
        });
        wrapper = shallow(<IconButton {...props} />);
      });
      it('should render the button with active state', () => {
        expect(wrapper).toHaveClassName('active round container-big');
      });
      it('should pass the isToggled prop to AccessibleButton', () => {
        expect(wrapper.find('AccessibleButton').prop('isToggled')).toBe(true);
      });
    });
    describe('when is not active', () => {
      let wrapper;
      beforeEach(() => {
        const props = createProps({
          isToggleButton: true,
          isToggled: false,
        });
        wrapper = shallow(<IconButton {...props} />);
      });
      it('should not render the button in the non active state', () => {
        expect(wrapper).not.toHaveClassName('active container-big');
      });
      it('should pass the isToggled prop to AccessibleButton', () => {
        expect(wrapper.find('AccessibleButton').prop('isToggled')).toBe(false);
      });
    });
    describe('when is disabled', () => {
      let wrapper;
      beforeEach(() => {
        const props = createProps({ isDisabled: true });
        wrapper = shallow(<IconButton {...props} />);
      });
      it('should render the button in the disabled state', () => {
        expect(wrapper).toHaveClassName('disabled round container-big');
      });
      it('should pass the isToggled prop to AccessibleButton', () => {
        expect(wrapper.find('AccessibleButton').prop('isDisabled')).toBe(true);
      });
    });
    describe('when is toggle', () => {
      let wrapper;
      beforeEach(() => {
        const props = createProps({
          isToggleButton: true,
          isToggled: false,
          theme: 'blue',
        });
        wrapper = shallow(<IconButton {...props} />);
      });
      it('should set the style according to the theme', () => {
        expect(wrapper).toHaveClassName(
          'default round container-big theme-blue'
        );
      });

      describe('when is active and disabled', () => {
        beforeEach(() => {
          const props = createProps({
            isToggleButton: true,
            isToggled: true,
            isDisabled: true,
            theme: 'green',
          });
          wrapper = shallow(<IconButton {...props} />);
        });

        it('should handle both classes', () => {
          expect(wrapper).toHaveClassName(
            'active disabled round container-big theme-green'
          );
        });
      });
    });
  });
});

describe('interaction', () => {
  describe('clicking the button', () => {
    let props;
    let wrapper;
    beforeEach(() => {
      props = createProps({ onClick: jest.fn() });
      wrapper = shallow(<IconButton {...props} />);
      wrapper.find('AccessibleButton').prop('onClick')();
    });
    it('should call the onClick callback', () => {
      expect(props.onClick).toHaveBeenCalledTimes(1);
    });
  });
});
