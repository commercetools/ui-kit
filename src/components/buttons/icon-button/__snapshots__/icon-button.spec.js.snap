// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering icon should output correct tree 1`] = `
<div
  className="button-appearance default round container-big theme-default"
  onMouseDown={[MockFunction]}
  onMouseOut={[MockFunction]}
  onMouseOver={[MockFunction]}
  onMouseUp={[MockFunction]}
>
  <AccessibleButton
    buttonAttributes={
      Object {
        "data-track-component": "IconButton",
      }
    }
    className="container-big"
    isToggleButton={false}
    label=""
    onClick={[Function]}
    type="button"
  >
    <div
      className="button-layout container-big"
    >
      <FilterIcon
        size="big"
      />
    </div>
  </AccessibleButton>
</div>
`;

exports[`rendering structure should output correct tree 1`] = `
<div
  className="button-appearance default round container-big theme-default"
  onMouseDown={[MockFunction]}
  onMouseOut={[MockFunction]}
  onMouseOver={[MockFunction]}
  onMouseUp={[MockFunction]}
>
  <AccessibleButton
    buttonAttributes={
      Object {
        "data-track-component": "IconButton",
      }
    }
    className="container-big"
    isToggleButton={false}
    label=""
    onClick={[Function]}
    type="button"
  >
    <div
      className="button-layout container-big"
    />
  </AccessibleButton>
</div>
`;
